libats-messaging-datatype/src/main/scala/com/advancedtelematic/libats/messaging_datatype/Messages.scala

```diff
@@ -159,7 +159,8 @@ object Messages {
       deviceUuid: DeviceId,
       result: InstallationResult,
       ecuReports: Map[EcuIdentifier, EcuInstallationReport],
-      rawReport: Option[String] = None
+      rawReport: Option[String] = None,
```

Comment:
Is not raw report a custom installation report?

Reply:
It's not
In example of DeviceUpdateCompleted which sends from client raw report is:
`"raw_report": "Installation succesful"`

Reply:
not sure I get it. raw_report is supposed to come from device and is opaque for backend. Is it read from different metadata field, means device is sending `raw_report` AND `installation_report`?

Reply:
Below you can find an example of DeviceManifestEcuSigned that client sends to Director
`custom` field contains `installation_report` which is custom installation report that should be send to Ericsson 
`raw_report` is not a installation report, it's just some additional data which sends by client
```
{
  "custom": {
    "installationReport": {
      "ecuAddress": "",
      "idBlocks": [
        {
          "downloadDate": "Wed Mar 11 13:41:01 2020\n",
          "downloadStatus": "SUCCESS",
          "installmentDate": "Wed Mar 11 13:41:09 2020\n",
          "installmentStatus": "SUCCESS",
          "packageName": "ivi-update",
          "softwareVersion": "0.0.4"
        }
      ],
      "model": "model",
      "vin": "0000000000000000"
    }
  },
  "ecu_version_manifests": {
    "2e7dfd1a327f6b892c4316d807c0875f464fad780cb2503f02c7ad8c7d1a084a": {
      "signatures": [
        {
          "keyid": "2e7dfd1a327f6b892c4316d807c0875f464fad780cb2503f02c7ad8c7d1a084a",
          "method": "rsassa-pss",
          "sig": "AB91j9DV6ELEe3C6+P/tLpzVWqVyW7TgYnvP2WOEt4mV9WcniSDnAeZ2MKt8DFnz7iiNyZraBaZHKj/Fciby/Imi/3Az0ZKpGKr8tlSXRaffLt9wdfRpM80UwX5wgPWRneK0mYknwf9WI5fNKhpOuWPQWF/oFFdZ2h8XFSmU9so72UGDd6CvK/voLcdrp/Ojh4FudYPltg8BxW154+13qhpeH8Z/urNUUdwaXxd9MvZrpfn0oXR3CzEZ9Mws75OMlrWd+HpYjjVYnWr2/5UWzbM+NztNwKAV7adNikRqBci1oNZmvzTWgfY0S2gGY9Ab5CusCkQ2GPrqBHP1jgdQTA=="
        }
      ],
      "signed": {
        "attacks_detected": "",
        "ecu_serial": "2e7dfd1a327f6b892c4316d807c0875f464fad780cb2503f02c7ad8c7d1a084a",
        "installed_image": {
          "fileinfo": {
            "hashes": {
              "sha256": "fd0d98e5e2c0765c4f597f36f46abd2579e3887388728427ac63b8aa1c4a3f38"
            },
            "length": 50331648
          },
          "filepath": "ivi-update_0.0.4"
        },
        "previous_timeserver_time": "1970-01-01T00:00:00Z",
        "timeserver_time": "1970-01-01T00:00:00Z"
      }
    },
    "e0e5e83fe0f6a7b6cec96123e1bff3e9c812d9cf1920b58f6ffd3ebc6429ec69": {
      "signatures": [
        {
          "keyid": "e0e5e83fe0f6a7b6cec96123e1bff3e9c812d9cf1920b58f6ffd3ebc6429ec69",
          "method": "rsassa-pss",
          "sig": "gyQ20uZBTyJhoCuF2NdZd1G2DDOBrMywfYrHtEJX9o81eUBAj8a0YKffpSVugcHr7crSxug/UIdeYvGkwvkftsCnl+sYZwYeHdhaQTwczNrVjf15EdKC4doYA6SOilbQ7DUICBnIAXQvEWxpwxb8P7JbuqEebRtlBdeegZDa9Ivjj2zM6SPZYzhl7xpCsiPmhoR8LA1shX4/WeasAapl8q59BgJPnZGkFJcjpACdVTF3qXEDEDDgZhxd+MYcy2oVXI1SkFvL7yOxAoC4ouY3l/cwMCyFPxdTAufoAtaD+Al02ec9yVqipaeFhmCy2CDColfmY0gmJpC9wt5jQxD6/Q=="
        }
      ],
      "signed": {
        "attacks_detected": "",
        "ecu_serial": "e0e5e83fe0f6a7b6cec96123e1bff3e9c812d9cf1920b58f6ffd3ebc6429ec69",
        "installed_image": {
          "fileinfo": {
            "hashes": {
              "sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
            },
            "length": 0
          },
          "filepath": "unknown"
        },
        "previous_timeserver_time": "1970-01-01T00:00:00Z",
        "report_counter": "6",
        "timeserver_time": "1970-01-01T00:00:00Z"
      }
    }
  },
  "installation_report": {
    "content_type": "application/vnd.com.here.otac.installationReport.v1",
    "report": {
      "correlation_id": "urn:here-ota:mtu:1ff64792-1c07-4f41-a2c4-a67f44cf7d57",
      "items": [
        {
          "ecu": "2e7dfd1a327f6b892c4316d807c0875f464fad780cb2503f02c7ad8c7d1a084a",
          "result": {
            "code": "OK",
            "description": "",
            "success": true
          }
        }
      ],
      "raw_report": "Installation succesful",
      "result": {
        "code": "OK",
        "description": "Device has been successfully installed",
        "success": true
      }
    }
  },
  "primary_ecu_serial": "e0e5e83fe0f6a7b6cec96123e1bff3e9c812d9cf1920b58f6ffd3ebc6429ec69"
}
```

Reply:
`raw_report` was introduced to send custom installation reports (I have double-checked with Laurent and Evgenii).

Reply:
sorry wrote that comment after I read that. So then we should just use raw_report?

Reply:
@simao yes. We probably should check if aktualizr's API wires the custom report correctly, but it should be `raw_report`

Reply:
Should we also parse it as json?

Reply:
I would suggest we do. It probably will not be a json in some cases, but then it can be packaged in something like 
```
{
  "report": "<base64>"
}
```

Reply:
an `aktualizr` does not provide an API to manipulate with the `raw_report`, this field is intended to use internally by the aktualizr, and `libaktualizr` has predefined string constants for different states of the installation result. So currently the aktualizr interprets raw_report as a string without any encoding and the `libaktualizr's` users have no access to it.

If I understand correctly the initial idea this field "raw_report" is intended to be used as a string or even as a list of strings separated by special characters like '\n' etc. 

Without any structure, correct me if I wrong!

Our custom "legacy installation report" has a predefined structure and can't be represented as a raw string. 

For this reason, we have a custom object in our manifest, this object is intended to be a container for any, not predefined objects in `libaktualizr`, serialized data which libaktualizr's users would like to send to BE in additional.

This is an example part of the manifest with the installation report.

```
{
  "installation_report": {
    "content_type": "application/vnd.com.here.otac.installationReport.v1",
    "report": {
      "correlation_id": "urn:here-ota:mtu:93e56a76-b570-44cf-adf0-ec8bcf3e5cb2",
      "items": [
        {
          "ecu": "84e1a345e035ca56fab7071c2ec141a04b193e1b05db3d12db613d38bf3fff57",
          "result": {
            "code": "OK",
            "description": "ANGEL-DELIGHT: Installing package was successful (YOU can add any additional information about the installation result HERE!!!!)",
            "success": true
          }
        }
      ],
      "raw_report": "Installation succesful",
      "result": {
        "code": "OK",
        "description": "Device has been successfully installed",
        "success": true
      }
    }
  }
}
```

@koshelev maybe you mean `raw_report` instead of `report`?

Reply:
> @patrickvacek Not sure who in the client team knows about this, but why is that data coming in `custom` and not `raw_report`?

Do you mean the content of the custom object?
If yes, we actually do not know in advance of its content, but we know that `custom: {}` exist and the client sends it anyway even if it's empty.

So from my understanding, the OTA BE should not worry about its contents of the `custo: {}` field, just redirect it as a `JSON` to the third party service, as it is, without parsing its content.

Reply:
> this field is intended to use internally by the aktualizr

wrong. The field is intended to be used exactly for the purpose of passing the opaque customer specific data. As mentioned above, if aktualizr is not using the field atm it will need to be changed.

> If I understand correctly the initial idea this field "raw_report" is intended to be used as a string or even as a list of strings separated by special characters like '\n' etc.

it is not correct. The initial idea was to enable a consumer of the API to pass any data it needs to the backend. Core product will not look into it and do not care how it is structured.

> Our custom "legacy installation report" has a predefined structure and can't be represented as a raw string.

actually, everything can be represented as a string - base64.

Reply:
> wrong. The field is intended to be used exactly for the purpose of passing the opaque customer specific data. As mentioned above, if aktualizr is not using the field atm it will need to be changed.

definitely we need some clarification on that! 
Do we have any description of the structure for the metadata?

Reply:
> actually, everything can be represented as a string - base64.

I agree it's needed if we are going to use `raw_report` field to send object serialized not in the JSON but as ASN.1 for example, or any other serialization mechanism.